edk2toollib/uefi/edk2/build_objects/dsc.py

```diff
@@ -0,0 +1,486 @@
+# @file dsc.py
+# Data model for the EDK II DSC
+#
+# Copyright (c) Microsoft Corporation
+#
+# SPDX-License-Identifier: BSD-2-Clause-Patent
+##
+
+# There will be some overlap between the objects for DSC files and FDF files
+import collections
+DEFAULT_SECTION_TYPE = "COMMON"
+
+
+class dsc_set(set):
+
+    def __init__(self, allowed_classes=[]):
+        self._allowed_classes = set(allowed_classes)
+
+    def add(self, item):
+        if len(self._allowed_classes) > 0 and type(item) not in self._allowed_classes:
+            raise ValueError(f"Cannot add {type(item)} to restricted set: {self._allowed_classes}")
+        if item in self:
+            super().discard(item)
+            # TODO: get add the old_item to item
+        super().add(item)
+
+
+class dsc_dict(collections.OrderedDict):
+    ''' A dictionary that allows specific classes as headers and sections '''
+
+    def __init__(self, allowed_key_classes=[], allowed_value_classes=[]):
+        self._allowed_key_classes = set(allowed_key_classes)
+        self._allowed_value_classes = set(allowed_value_classes)
+
+    def __missing__(self, key):
+        if len(self._allowed_value_classes) > 0:  # if we have specified allowed value classes, make a new dsc_set
+            self[key] = dsc_set(self._allowed_value_classes)
+            return self[key]
+        raise KeyError(key)
+
+    def __setitem__(self, key, val):
+        if len(self._allowed_key_classes) > 0 and type(key) not in self._allowed_key_classes:
+            raise ValueError(f"Cannot add {type(key)} to restricted set: {self._allowed_key_classes}")
+
+        if len(self._allowed_value_classes) > 0:
```

Comment:
I think the order of items matters because there is the case that the later item override the former one.  I think the val should be a list,  dsc_list?

Reply:
So I'm not sure I understand the comment. You're saying we need to keep track of the items in the order that they come in. This code just restricts specific the type of items that can be added to the specific object. But this uses a ordered dictionary, so it keeps track of the order that they were added.

Reply:
Actually I think I understand now. We have a set that doesn't keep ordering. I see what you're saying.

Reply:
Added this as an issue #82